CONTRIBUTING.md

```diff
@@ -345,6 +345,25 @@ with clear instructions on how to install the required libraries.
 Furthermore, if your package requires CGO, it MUST be able to compile and operate in a no op mode
 or report a warning back to the collector with a clear error saying CGO is required to work.
 
+### Breaking changes
+
+Whenever possible, we adhere to semver as our minimum standards. Even before v1, we strive to not break compatibility
+without a good reason. Hence, when a change is known to cause a breaking change, it should be clearly marked in the
+changelog, possibly with a line instructing users how to move forward.
+
+We also strive to perform breaking changes in two stages: 
+
+- when we need to remove something, we mark a feature as deprecated in one version, and remove it in the next one
+- when renaming types, functions, or attributes, we create the new name first, deprecating the old one in one
+version, removing it in the next version
+- when creating new interfaces, provide helpers so that consumers can be forward compatible without much effort. One
```

Comment:
Interfaces (generally) [cannot be added to](https://pkg.go.dev/golang.org/x/exp@v0.0.0-20220114162006-9d54fb35363c/apidiff#readme-interfaces) in a backward compatible manner and I think we should avoid contortions to try to give the appearance that they can.  I also think that we should be [very careful about adding unexported methods](https://github.com/open-telemetry/opentelemetry-collector/issues/4674) to interfaces in an attempt to reserve the ability to add to them later.

The `ServerAuthenticator` helper seems reasonable in that it provides sane default behavior while not being the only way to implement the interface, but some of it is now unnecessary since [nil `StartFunc` and `ShutdownFunc`](https://github.com/open-telemetry/opentelemetry-collector/blob/6b5a3d08a96bfb41a5e121b34f592a1d5c6e0435/component/componenthelper/component.go#L24) have that sane default behavior and an authenticator with default AuthN functionality is useless.  If we were creating this interface today, would we still want this helper or would the existing ways of implementing an `Extension` be sufficient?

I don't think this is the place to discuss the merits of that particular interface, but I do think it might be the case that guidance for how to handle interface compatibility is more complicated and should be discussed and added separately from the rest of this guidance which is more straightforward.

Reply:
I agree with you, but in practical terms, when developing an authenticator, I might not care whether my authenticator is a `ServerAuthenticator` or a `ValidatableServerAuthenticator`. I'd prefer to only provide an `Authenticate` function.

Packages like `confighttp` and `configgrpc` might care about the validation aspect of the component, and `configauth` might provide a default `Validate` that complies with that new interface. IMO, this is the value of providing those helpers: new interfaces can be composed and used in consuming places without having to adapt all the current implementations of the simpler interface.

Reply:
In the case of validation I assume the default behavior if no validator is provided is to pass validation.  That could also be achieved by testing for an implementation of a `Validater` interface and only performing validation if it is implemented.  This puts the burden on the consumer, who is in the best position to know whether the new capability is optional (thus taking the interface testing approach) or mandatory (and thus taking an interface composition approach and requiring a `ValidatableServerAuthenticator`).

Reply:
I am strongly opposed to this being a `MUST`, as it is in the current revision.  I think if the discussion of interfaces were removed from this PR and handled separately that I could approve this PR, but as it stands I cannot.

Reply:
Alright, I just removed it from here. We can discuss about it and add a better version in the future.